tools/tpm2_policycountertimer.c

```diff
@@ -0,0 +1,251 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+
+#include <stdlib.h>
+
+#include "files.h"
+#include "tpm2.h"
+#include "tpm2_nv_util.h"
+#include "tpm2_options.h"
+#include "tpm2_policy.h"
+
+typedef struct tpm_policycountertimer_ctx tpm_policycountertimer_ctx;
+struct tpm_policycountertimer_ctx {
+    //Inputs
+    const char *session_path;
+    tpm2_session *session;
+
+    TPM2B_OPERAND operand_b;
+
+    uint16_t offset;
+
+    TPM2_EO operation;
+
+    //Output
+    const char *policy_digest_path;
+};
+
+static tpm_policycountertimer_ctx ctx = {
+    .operation = TPM2_EO_EQ,
+    .offset = 0,
+};
+
+static bool process_time_info(char *value, uint16_t offset, uint8_t size) {
+
+    ctx.offset = offset;
+
+    uint64_t datatype64 = 0;
+    uint32_t datatype32 = 0;
+    bool result = true;
+
+    char *string_bytes = malloc( 1 + (size * 2) );
+    if (!string_bytes) {
+        LOG_ERR("oom");
+        return false;
+    }
+
+    switch(size) {
+        case sizeof(uint64_t):
+            result = tpm2_util_string_to_uint64(value, &datatype64);
+            if (!result) {
+                LOG_ERR("Couldn't convert starting offset, got: \"%s\"", value);
+                goto process_time_info_out;
+            }
+            snprintf(string_bytes, (1 + (size * 2)), "%016lX", datatype64);
+            break;
+        case sizeof(uint32_t):
+            result = tpm2_util_string_to_uint32(value, &datatype32);
+            if (!result) {
+                LOG_ERR("Couldn't convert starting offset, got: \"%s\"", value);
+                goto process_time_info_out;
+            }
+            snprintf(string_bytes, (1 + (size * 2)), "%08X", datatype32);
+            break;
+    }
+
+    ctx.operand_b.size = size;
+    int rc = tpm2_util_hex_to_byte_structure(string_bytes, &ctx.operand_b.size,
+        ctx.operand_b.buffer);
+    if (rc) {
+        LOG_ERR("Error converting the milliseconds input.");
+        result = false;
+    }
```

Comment:
I have almost no idea what you're trying to do here and this code looks very wrong.

You have some data structure you need to fill up with a number starting as a string right? The data structure is either 64 or 32 bits wide? Is this your "to big endian" routine?

Cant you do?
```C
if (is_32_bit) {
  tpm2_util_string_to_uint32(str, value);
  tpm_structure.field = tpm2_util_hton_32(str, value);
} else {
  tpm2_util_string_to_uint64(str, value);
  tpm_structure.field = tpm2_util_hton_64(str, value);
}
```

Reply:
> I have almost no idea what you're trying to do here and this code looks very wrong.
> 
> You have some data structure you need to fill up with a number starting as a string right? The data structure is either 64 or 32 bits wide? Is this your "to big endian" routine?
> 
> Cant you do?
> 
> ```c
> if (is_32_bit) {
>   tpm2_util_string_to_uint32(str, value);
>   tpm_structure.field = tpm2_util_hton_32(str, value);
> } else {
>   tpm2_util_string_to_uint64(str, value);
>   tpm_structure.field = tpm2_util_hton_64(str, value);
> }
> ```

No. We are not filling a data structure. Rather we take the number starting as a string, convert it to hex and then break it into a byte stream which forms the operand buffer. The operand buffer is not a single specific structure member but could be any of the structure members merely used for comparison. And so an offset is additionally specified in the TPM/ESAPI command. The TPM then consumes the offset and the buffer for its internal comparison of the buffer and a corresponding TPM state/value.

Reply:
I wish the TPM command itself was simpler - rather than have a buffer and structure member offset as an index into the structure members, why not have unique inputs for all structure members in which case your proposed routine would have been handy.

Reply:
Libmu have anything?

Reply:
So you're convert logic should be something like:
1. Get the key and value the user wants to set.
2. Convert the value the user provides into either a u64 or u32.
3. serialize that to the TPM2B_OPERAND
4. set offset for the command to the value of offsetof() macro for the key field in the TPMS structure.

For step 2, you're converting it to a hex string and and then populating that hex string in into the TPM2B_OEPRAND buffer. That should be equivalent of converting it to a big endian and populating the buffer. Just call  the host to tpm2_util_hton_32 or tpm2_util_hton_64 routines, and then copy that into the buffer as is. No need to be malloc'ing up.